sockets: Properly refcount TCP sockets

pfalcon opened the pull request:
net: socket: Increment TCP context refcount on socket(), accept()
    
    TCP context is effectively owned by both application and the stack:
    stack may detect that peer closed/aborted connection, but it must
    not dispose of the context behind the application back. Likewise,
    when application "closes" context, it's not disposed of immediately,
    there's yet closing handshake for stack to perform.
    
    This effectively means that TCP contexts have refcount of 2 when
    they're created. Without this change, following situation is
    possible: peer opens connection, an app get a context (or socket)
    via accept, peer sends data, closes connection. An app still holds
    a reference to connection, but stack may dispose of context, and
    even reuse it for a new connection. Then application holds a reference
    to either free, or completely different context.
    
    This situation was very clearly and 100% reproducible when making
    Zephyr port of open62541 library, which works in async manner using
    select().
    
    Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: socket: Increment TCP context refcount on socket(), accept()

TCP context is effectively owned by both application and the stack:
stack may detect that peer closed/aborted connection, but it must
not dispose of the context behind the application back. Likewise,
when application "closes" context, it's not disposed of immediately,
there's yet closing handshake for stack to perform.

This effectively means that TCP contexts have refcount of 2 when
they're created. Without this change, following situation is
possible: peer opens connection, an app get a context (or socket)
via accept, peer sends data, closes connection. An app still holds
a reference to connection, but stack may dispose of context, and
even reuse it for a new connection. Then application holds a reference
to either free, or completely different context.

This situation was very clearly and 100% reproducible when making
Zephyr port of open62541 library, which works in async manner using
select().

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
This fixes very old known issue, mentioned and attempted to work around before, e.g. https://github.com/zephyrproject-rtos/zephyr/pull/7865

It expose itself as 100% blocker while porting https://github.com/open62541/open62541 - there, everything was as described in the commit message: open62541 mainloop had an old socket in select(). In the meantime, it was closed and a new conn opened, getting the same context. At which point select() woke up, figure that the old connection is closed, but issues close() on the new connection, which led to obviously weird behavior.

Now, I initially fixed it on socket level, and tested that it can handle 1000 consecutive connections (which should answer a question "but what if such increment leads to obvious context leaks?" - obvious, not; but as we know, we have a lot of corner cases in the stack).

Today, when brushing up that code for submission, I asked @rlubos if he thinks it would make sense to move that stuff down to net_context level (because otherwise I'm pretty tired of this stuff, most of which is however not socket issues, but POSIX subsys vs newlib header conflicts). He said yes, and tried to move net_context_ref() calls to corresponding net_context functions, but no matter how simple and obvious that, it didn't behave as expected, specifically deadlocked after 6th connection.

So, this PR includes net_context-level attempt, then it's revert, then socket-level patch which works (verified again with 1000 connections). See if you guys can spot an obvious thinko with net_context attempt, but otherwise we'll just drop it and stick with socket-level one.

pfalcon committed:
net: context: put: Update for shared TCP context ownership

TCP context is now created with refcount of 2, signifying that it's
jointly owned by an app and stack. Thus, net_context_put()
unconditionally calls net_context_unref() to decrement refcount on
app's behalf, and leaves stack's refcount to internal routines
which handle sending/receiving/timing out FINs, etc.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

jukkar commented:
Is there some rebase issue in this PR as I see a commit submitted, then reverted and then re-submitted again?

pfalcon commented:
> Is there some rebase issue in this PR as I see a commit submitted, then reverted and then re-submitted again?

No, it's not a rebase issue, the matter is discussed in comment above, https://github.com/zephyrproject-rtos/zephyr/pull/13667#issuecomment-466384694

pfalcon commented:
Ok, I assume nobody has an idea what may be wrong with first attempt at change, so I just removing it/its revert from this patch, and leaving only what's known to work.

@jukkar, @tbursztyka, @rlubos: Please review, this is a serious issue, with it not fixed it doesn't make sense to perform release-wise testing of networking functionality.

rlubos commented:
@pfalcon Do you know if there is a way to reproduce this issue with existing samples? I find it really awkward that it actually works when you increse refcount at the socket level, but not when you do it at net_context level.

Any chance to have the original attempt reuploaded (on a separate branch perhaps, to keep this PR clear)? I did not have a chance earlier to analyze how actually this ref counting of net_context work and provide input.

pfalcon commented:
> @pfalcon Do you know if there is a way to reproduce this issue with existing samples?

No beyond that pretty nondererministic TCP recv window error. Or it would have been fixed long ago ;-). But it can be easily reproduced with http://github.com/pfalcon/open62541demos if you'd like to give it a try ;-). (Well, as a full disclosure, it was easily reproducible before https://github.com/zephyrproject-rtos/zephyr/pull/13230, I did this patch likely before it, just submit in different order).

> I find it really awkward that it actually works when you increse refcount at the socket level, but not when you do it at net_context level.

That's why I posted initial version with that patch included, so interested parties could eyeball it. Possibly some exit case wasn't accounted for or something. Well, I assume that everybody already burnt out by this release work and doesn't have mental energy to explore sideways, and we didn't even fully finished with net_buf refactor. So yeah, can just use what works.

> I did not have a chance earlier to analyze how actually this ref counting of net_context work and provide input.

It's still in https://github.com/zephyrproject-rtos/zephyr/pull/12984/commits branch (among last commits as of now).

rlubos commented:
Sorry for not doing it earlier, but to actually provide any (hopefully) valuable input, I needed to do my homework first and understed how this net_context reference/unreference works.

What I would find intuitive is that `net_context_get` would increase ref counter once for the application (and this one should be freed with `net_context_put`), and then, if needed, bump the ref counter for the underlaying protocol, if it needs to manage it. 

So looking at the net_context patch (https://github.com/zephyrproject-rtos/zephyr/commit/218e010ac1141df11972dd3ab12bb5bc0e2da5e0#diff-a278935f8d5d5e0b4424ca3e1182d48cR2435) I find it suspicious to increment ref counter within `NET_CONN_CB(tcp_syn_rcvd)`. My understanding is that we already did this for TCP within `net_context_get`, which we call just before to grab a new context. So now we have ref counter increased three times, yet we unref it only twice. Unless I miss something of course.

pfalcon commented:
> Sorry for not doing it earlier, but to actually provide any (hopefully) valuable input, I needed to do my homework first and understand how this net_context reference/unreference works.

NP, and well, time goes forward, and I'd like to start integration testing for the release, and there's huge difference between testing master rev X, vs master rev X + random patches U, V, W on top of that. The latter leads to mistakes and non-reproducibility. So, while I'd like to get to the bottom of it, eager to get stuff into master too.

> My understanding is that we already did this for TCP within `net_context_get`, which we call just before to grab a new context. So now we have ref counter increased three times, yet we unref it only twice. Unless I miss something of course.

D'oh, indeed, you found the bug. But now I'm not sure how to fix it. The accept (i.e. NET_CONN_CB(tcp_syn_rcvd)) case has a number of error exits. What to do with them - call net_context_unref() twice? Sounds hackish. I can propose to add net_context_get_internal(), which would increment it once, then make net_context_get() which would increment it again for the TCP case. accept() would also call net_context_get_internal(). Effectively, we'd extra increment refcount only when we guaranteedly handover a context to user.

That sounds correct, but... it's more complicated than doing it on socket level like this patch does ;-).

rlubos commented:
Well, what I'm actually missing in the "accept" is `net_context_put`. Current implementation appears to mix the concept of context allocation and context referencing. We've allocated the context with `net_context_get` and added extra reference with `net_context_ref` - we need to unwind that with `net_context_unref` and `net_context_put` in case of error, what's hacky in that? Especially that this could be easily handled by an extra goto label at the bottom of `NET_CONN_CB(tcp_syn_rcvd))`. 

What I don't like with current proposal is that it actually mixes the two approaches - one one hand we bump the ref counter at socket level, ont the other we unref it at `net_context` level. And it gives me the impression of ad-hoc patching of TCP issues at socket level (but we actually do the same at net_context patch) . That said, for even neater approach, we should perhaps bump the ref counter within `net_tcp_get`, as all in all, it's TCP implementation that should unref it in the end. This way, `net_context` will only need to care about the "application" ref count increase/decrease.

pfalcon commented:
> Current implementation appears to mix the concept of context allocation and context referencing. We've allocated the context with `net_context_get` and added extra reference with `net_context_ref` - we need to unwind that with `net_context_unref` and `net_context_put` in case of error, what's hacky in that?

So, what's hacky is that there should be no net_context_put(), there should be just net_context_unref(). Currently, it's unclear which does what, and how they differ, and use of one or another is dictated by trial and error (YMMV).

> That said, for even neater approach, we should perhaps bump the ref counter within `net_tcp_get`

Yes, that's even neater and layered, except the situation with decrementing is much less clear, and all that will require even more changes, and even more debugging, as likely they won't work from the start, because we don't really have that ideal layeredness.

All in all, I'd like to hear @jukkar's opinion of what of that of that is worth to be pursued. I'm happy to look into that, but nor for outcome like https://github.com/zephyrproject-rtos/zephyr/pull/13425

pfalcon commented:
@jukkar: Ping

jukkar commented:
> What I don't like with current proposal is that it actually mixes the two approaches - one one hand we bump the ref counter at socket level, ont the other we unref it at `net_context` level.

That is indeed bad, (mixing ref and unref components) and it will be very difficult to understand it later. So if net_context increments refcount, it should be responsible to unref it too, if this cannot be done then at least this exception needs to be documented very well.

By just looking to the code in this PR, it is very difficult to see what it is trying to accomplish. So it would be really helpful if we could have a test program (other than porting some 3rd party app to zephyr) that would trigger the issue, would it be possible to create one? Then it is just a matter to enabling debug level / using debugger to see how the refcount is behaving.

pfalcon commented:
> By just looking to the code in this PR, it is very difficult to see what it is trying to accomplish.

This PR has more comments than code, and has a detailed commit message. What's exactly unclear in "TCP context is effectively owned by both application and the stack" ?

> So it would be really helpful if we could have a test program (other than porting some 3rd party app to zephyr) that would trigger the issue, would it be possible to create one?

I don't know how to create one. Our stack isn't exactly designed for modularity, edge condition injection, and testing. So, I suggest to start with understanding the issue (which is very simple again), and then just find the best place to address it. The patch presented here is however resolves the issue as it is.

pfalcon commented:
> I don't know how to create one.

Actually, I can sit and try. I won't be able to reproduce the epic case when an app waited for events on a socket, which in the meantime had its net_context freed by the stack. Then app woke up, git EOF event, and closed the socket, but that net_context already had a new connection. - But well, I can at least try to reproduce *something*. Will work on it tomorrow.

jukkar commented:
> This PR has more comments than code, and has a detailed commit message. What's exactly unclear in "TCP context is effectively owned by both application and the stack" ?

What I meant is that by looking the diff, it is quite hard to understand all the implications there might be to other parts of the system. Comments are good but if it would be best if the could test this by running a piece of code.

pfalcon commented:
It's actually quite easy to reproduce that case, as soon as we: 1) actually test handling of EOF on socket, like https://github.com/zephyrproject-rtos/zephyr/pull/13995 does; 2) test that operations happen on non-freed net_context (will post trivial patch later).

pfalcon committed:
net: sockets: recv_stream: Check that the underlying net_context active

It may be closed by the stack behind our back (something which needs
to be fixed).

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
> 2) test that operations happen on non-freed net_context

Posted as https://github.com/zephyrproject-rtos/zephyr/pull/14050

@jukkar, Please review.

pfalcon commented:
Ok, so we had a TDD case, when the check exposing the issue was merged sooner than the fix, and broke master. I resubmit it again to demonstrate the issue: https://github.com/zephyrproject-rtos/zephyr/pull/14136 . I then include it here, and hope we'll be done with it.

jukkar reviewed (approved):
LGTM

pfalcon commented:
Ping @rlubos

rlubos commented:
I've made my opinion here - these are `net_context`/`tcp` compablity issues and IMO should be fixed at these layers.

I'm not going to block this approach to fix the problem, but don't expect me to accept it either.

pfalcon commented:
@tbursztyka: Please review.

This PR fixes a serious bug in the IP stack. Without it, network sockets should be considered completely broken. Thus, this PR should be considered high-priority for the release. Adding "TSC" label so this doesn't slip past the attention.
